src/language/autoComplete.ts

```diff
@@ -0,0 +1,216 @@
+import { languages, IRange } from 'monaco-editor/esm/vs/editor/editor.api';
+import { MACROS } from './macros';
+import { Monaco } from '@grafana/ui';
+
+interface KeyValue {
+  [key: string]: string[];
+}
+
+interface SuggestionDefinition extends Omit<languages.CompletionItem, 'range' | 'insertText'> {
+  insertText?: languages.CompletionItem['insertText'];
+}
+
+enum SuggestionType {
+  'all',
+  'macros',
+  'tables',
+  'fields',
+}
+
+const tableColumns: KeyValue = {
+  asset: ['asset_id', 'asset_name', 'asset_description', 'asset_model_id'],
+  asset_property: ['property_id', 'asset_id', 'property_name', 'property_alias', 'asset_composite_model_id'],
+  raw_time_series: [
+    'asset_id',
+    'property_id',
+    'property_alias',
+    'event_timestamp',
+    'quality',
+    'boolean_value',
+    'int_value',
+    'double_value',
+    'string_value',
+  ],
+  latest_value_time_series: [
+    'asset_id',
+    'property_id',
+    'property_alias',
+    'event_timestamp',
+    'quality',
+    'boolean_value',
+    'int_value',
+    'double_value',
+    'string_value',
+  ],
+  precomputed_aggregates: [
+    'asset_id',
+    'property_id',
+    'property_alias',
+    'event_timestamp',
+    'resolution',
+    'sum_value',
+    'count_value',
+    'average_value',
+    'maximum_value',
+    'minimum_value',
+    'stdev_value',
+  ],
+};
+
+interface SitewiseCompletionProviderType extends languages.CompletionItemProvider {
+  fetchSuggestions(range: IRange, types: SuggestionType, table: null | string): languages.CompletionItem[];
+  tableDefinitions(): SuggestionDefinition[];
+  fieldDefinitions(table: string): SuggestionDefinition[];
+  macroDefinitions(range: IRange): SuggestionDefinition[];
+  allDefinitions(range: IRange, table: null | string): SuggestionDefinition[];
+  buildAutocompleteSuggestion(definition: SuggestionDefinition, range: IRange): languages.CompletionItem;
+  monaco: null | Monaco;
+  currentSpace: string;
+}
+
+// TODO: Check out getStandardSQLCompletionProvider to get standard SQL completion
```

Comment:
I think looking at https://github.com/grafana/plugin-ui/tree/main/src/components/SQLEditor might be helpful. There's some useful tools there such as `LinkedToken` that make it easier to build up an autocomplete feature instead of building up an array of the tokens like it's done below.

Reply:
Can definitely look at it, but let's do it as a separate PR?